drivers: disk_flash: Create disk_flash driver

kica-z opened the pull request:
For our project we created a driver that emulates a flash on an underlying disk. We use this to emulate the secondary partition of mcuboot on an spi SD-Card. Since this driver could be usefull to some people we would like to upstream it.

Before that there are however some questions I have since this is the first zephyr driver I develop:
- Is this driver a welcome addition to the codebase or is there no use case outside our project?
- should the `init_level` of the driver be configurable? Currently it is `APPLICATION` since the `ramdisk` driver is initialized at `APPLICATION` level. If yes how should I go about that? Add it to the devicetree?
- Can the `FLASH_DISK_BUFFER_SIZE` be evaluated at compile time from the underlying disk? Or is there a way to determine and allocate it during runtime without having to define a Kconfig?
- Is there a nicer way to define the device init macro than to define it twice just because of the `flash_pages`?

Laczen commented:
@kica-z, thanks for the PR this feature would indeed be highly appreciated.

One difficulty I see: the proposed PR is kind of making a partition on disk by specifying the offset and the size. But the other systems using the disk (fatfs, littlefs) are not aware that this is happening, how is this handled ?

kica-z commented:
@carlescufi, @de-nordic I moved the disk-sector size as well as the init-level of the driver into the dts so that it can be adjusted for each instance individually.

@de-nordic Regarding the SD mcuboot example I can try to add one but this will take some time since for our project we use the nordic tools and the nrf fork so some adjustment is necessary.

@de-nordic What do you mean when you say keeping in mind that the sd-card is removable. Should not access to the disk-api fail as soon as the sd-card is disconnected? What would you expect from the driver in this scenario? Wait for it to reapear and reinit with the new sector size? Or just check every call if the sector buffer is sufficient?

@Laczen Currently we in the project create a separate partition on the sd-card for this and manually take care that there is no overlap between the fs and the disk_flash. It is true, that when the user defines overlapping partitions with e.g. fatfs,littlefs this will not be detected and lead to errors. Do you think there has to be a mechanism to detect this instead of leaving this to the user?

Laczen commented:
> @Laczen Currently we in the project create a separate partition on the sd-card for this and manually take care that there is no overlap between the fs and the disk_flash. It is true, that when the user defines overlapping partitions with e.g. fatfs,littlefs this will not be detected and lead to errors. Do you think there has to be a mechanism to detect this instead of leaving this to the user?

I don't see how this can be done. The sd-card size is unknown (maybe this is what @de-nordic is referring to with the removable sd-card) and there is no partitioning of sd-cards (correct me if I'm wrong).

Laczen commented:
> @Laczen we do actually partition our sd-card. In our project fatfs uses the `D:` partition and the disk_flash is writing into the OEM partition. ![image](https://user-images.githubusercontent.com/126649875/267694248-9bc9b0b6-ea0c-49ff-b8cf-ba56875ecdb7.png)

What does zephyr report about this disk ? Can it detect the D partition and the OEM partition ? How is the fatfs limited to the `D:` partition ?

kica-z commented:
@Laczen The disk-api reports the full size of the sd-card no matter what partitions are defined on the card (calculated by sector_count * sector_size). Fatfs however detects the filesystem in the first partition and reports only the size of that partition as its size (f_blocks * f_frsize).

So at least in this constellation (fs partition first -> oem after) it seems to work as intended. The disk_flash driver of course still needs to be tuned to write into the oem partition as there is no autodetection or similar implemented here.

Laczen commented:
@kica-z it seems fragile to rely on the existence of a fatfs partition to allow this to work. At first glance this wouldn't be supported when littlefs is used on the disk.

kica-z commented:
@Laczen I am not quite sure I understood your comment correctly so correct me if i misinterpreted it:

I don't think the driver is responsible for supporting the user in configuring his block device.
Both fatfs and littlefs have mechanisms that allow them to not use the entire space on a disk. So if a user wants to use a FS in addition to this driver that is fine as long as it does not read/write on the same diskspace as the disk_flash. Even read/write on the same diskspace would be fine from disk_flash point of view but the FS will likely get corrupted.

So in conclusion the driver is not relying on a fatfs partition to be present. The user simply has to configure his setup according to his needs which might then need a fatfs or littlefs that is not using the entire diskspace of a disk.

I however agree that this might be a topic for documentation in order to support the user with using this driver in parallel to a fs.

Laczen commented:
> I don't think the driver is responsible for supporting the user in configuring his block device. Both fatfs and littlefs have mechanisms that allow them to not use the entire space on a disk. So if a user wants to use a FS in addition to this driver that is fine as long as it does not read/write on the same diskspace as the disk_flash. Even read/write on the same diskspace would be fine from disk_flash point of view but the FS will likely get corrupted.

I don't think this is true for littlefs, it always uses the entire disk. I also doubt that it is possible to create a fatfs partition in zephyr that uses only part of the disk, so if a mkfs is done in zephyr this will also consume the whole disk.

kica-z commented:
@Laczen I stand corrected littlefs really seems to not support partially using a disk. I misinterpreted a documentation entry. I don't know mkfs of zephyr well enough to make a statement regarding the possibility of not using the whole disk. It is quite possible that this is not possible.

However you could circumvent this problem by defining two disk_flash devices and running littlefs/fatfs on one of them as both FS also support the flash_api. This way there would be no overlap. I have however not tested this yet.

But as already mentioned I think the support of a filesystem on the same disk is not a requirement for this driver to be usefull. It would definitely be a nice to have though.

Laczen commented:
> However you could circumvent this problem by defining two disk_flash devices and running littlefs/fatfs on one of them as both FS also support the flash_api. This way there would be no overlap. I have however not tested this yet.
> 

This would indeed be a solution to avoid the problem. But I don't think it is possible at the moment to "split" or partition a disk. 

@danieldegrasse are there any plans to allow splitting or partitioning a disk (e.g. some fixed size raw zephyr-disk, and then a variable sized disk for a filesystem) ?

> But as already mentioned I think the support of a filesystem on the same disk is not a requirement for this driver to be usefull. It would definitely be a nice to have though.

I agree it would be nice to have this,

danieldegrasse commented:
> @danieldegrasse are there any plans to allow splitting or partitioning a disk (e.g. some fixed size raw zephyr-disk, and then a variable sized disk for a filesystem) ?

No plans currently from me. I expect this could be achieved by defining an emulated disk driver that simply reported a different sector count than the underlying disk device it was exposing. Multiple disk drivers like this could be combined to "parition" a disk with the current API

As far as the PR itself, this looks like a useful addition. Although the only use case I'm aware of is MCUBoot, there are still likely cases where users want to use MCUBoot as the bootloader but store boot images on a SD or eMMC device

kica-z commented:
> No plans currently from me. I expect this could be achieved by defining an emulated disk driver that simply reported a different sector count than the underlying disk device it was exposing. Multiple disk drivers like this could be combined to "parition" a disk with the current API

I thought about this too and think it could be a useful addition. Maybe even add a devicetree entry `partitions` for disk devices like it currently is available for flash via `fixed-partitions`. However I think this should be discussed in a separate `issue` or `PR`.

kica-z committed:
drivers: disk_flash: Create disk_flash driver

Create a driver that emulates a flash storage on the disk api.
This enables e.g. using an spi SD-Card as secondary partition
for mcuboot.

Signed-off-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>

tmon-nordic commented:
> @de-nordic What do you mean when you say keeping in mind that the sd-card is removable. Should not access to the disk-api fail as soon as the sd-card is disconnected? What would you expect from the driver in this scenario? Wait for it to reapear and reinit with the new sector size? Or just check every call if the sector buffer is sufficient?

Checking every call if the sector buffer is sufficient is not enough in my opinion. For removable drives, I think some sort of runtime ID value assigned with each drive insertion would make sense. If the calls are bound to the runtime ID then there's no risk of e.g. writing sector to completely different card. However, please treat this just as a consideration rather than an requirement because I have no clue how to sanely implement such mechanism myself.

kica-z commented:
@de-nordic Is there something that I have yet to do before a review can happen or will this be reviewed given some time?

danieldegrasse commented on drivers/flash/Kconfig.disk:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config FLASH_DRIVER_DISK
+	bool "Use the flash on disk emulator"
```
The convention for drivers is to define the Kconfig like so:

```
config FLASH_DRIVER_DISK
	default y
	depends on DT_HAS_ZEPHYR_DISK_FLASH_ENABLED
```

This way, the driver is enabled whenever the subsystem is available (`CONFIG_FLASH=y`) and an enabled node exists in the devicetree

kica-z replied:
I didn't know this. So this would be the prefered way? Or is the select not necessary?
```
config FLASH_DRIVER_DISK
	bool "Use the flash on disk emulator"
	default y
	depends on DT_HAS_ZEPHYR_DISK_FLASH_ENABLED
	select FLASH_HAS_DRIVER_ENABLED
	help
	  This option enables the flash on disk emulator
	  which can be used to emulate a flash on a disk
	  device like sd-card or ramdisk.
```

danieldegrasse commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+      offset on the disk in bytes. If the offset is 10 the 10th byte will
+      be the first one used.
+    default: 0
+  init-level:
```
I don't think this property or `init-priority` should be devicetree properties. Rather, we should use `POST_KERNEL` as the initialization level, and define a Kconfig `CONFIG_FLASH_DISK_INIT_PRIORITY` (if the init level needs to be different from `CONFIG_FLASH_INIT_PRIORITY`.)

kica-z replied:
I thought that initialy too but then I discovered, that RAM-Disk is initialised in level "APPLICATION". As this driver is reliant on the fact that the underlying disk has already been initialized both `init-level` and `init-priority` need to be adjustable per instance as the underlying disk can be different depending on the instance.

Hovewer please do tell me if you have another idea how to handle this.

jfischer-no replied:
> I thought that initialy too but then I discovered, that RAM-Disk is initialised in level "APPLICATION".

No, https://github.com/zephyrproject-rtos/zephyr/blob/e364f29bffdfb122fead4b7dad082428f2f6969c/drivers/disk/ramdisk.c#L169

kica-z replied:
@jfischer-no Fair enough it is not anymore but at the time i wrote the driver it was ([ramdisk.c](https://github.com/zephyrproject-rtos/zephyr/blob/a5fd0d184a15d0af52599e0f5f5af880f353c6d4/drivers/disk/ramdisk.c)). Therefore initialization was not possible in post-kernel now it is.

danieldegrasse commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+    description: >
+      The initialization priority of this driver. It is important, that the underlying disk has
+      already been initialized.
+  write-block-size:
```
Why would this need to be configurable? The underlying disk controller should always support 1 byte block writes

kica-z replied:
I just wanted to emulate a flash-disk as closely as possible. So your proposal would be to remove both parameter from DT and hardcode them into the flash_parameters?

danieldegrasse commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+    type: int
+    description: size of a flash write block in bytes
+    default: 1
+  erase-value:
```
Same question- why would a user need to configure the erase value here?

kica-z replied:
My thought was that a user might want a defined value in the memory when it is deleted. Would you propose to just set this to 0?

danieldegrasse commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+			      &disk_flash_config_##index, \
+			      DT_STRING_TOKEN(DT_DRV_INST(index), init_level), \
+			      DT_INST_PROP(index, init_priority), &disk_flash_api);
+/* clang-format on */
```
Is this comment needed?

kica-z replied:
Well it is not really needed but if clang-format is not set to `off` the whole `define` gets formatted in a really weird way.

carlescufi commented:
@tmon-nordic @jfischer-no @danieldegrasse could you please take a look?

jfischer-no commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+      The initialization level of this driver. It is important, that the underlying disk has
+      already been initialized.
+    default: "APPLICATION"
+  init-priority:
```
Cannot be devicetree property.

kica-z replied:
Why? I understand that with the new ramdisk cleanup it is now initialized post kernel but should the user not have the freedom to adjust this to his needs?

jfischer-no replied:
Because it is always a Kconfig option, here for flash drivers FLASH_INIT_PRIORITY.

kica-z replied:
I replaced this with the `FLASH_INIT_PRIORITY`.
-> I still think this is less flexible but it works. It has to be overwritten though since `ramdisk` has the same priority

jfischer-no commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+			(.flash_parameters = \
+				{ \
+					.write_block_size = DT_INST_PROP(index, write_block_size), \
+					.erase_value = DT_INST_PROP(index, erase_value) \
```
`erase_value = DT_INST_PROP_OR(i, erase_value, 0xFF),`

kica-z replied:
Removed the DT property

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+{
+	LOG_DBG("%s called", __func__);
+	struct flash_disk_dev_data *data = dev->data;
+	const struct flash_disk_dev_config *cfg = dev->config;
```
I adjusted this accordingly.

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+	}
+}
+
+/* clang-format off */
```
Removed

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+		.buf = buf_##index, \
+		.bufsize = sizeof(buf_##index), \
+	}; \
+\
```
I reworked the macro a bit. If there is still room for improvement in formatting please advise what could be better.

kica-z commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: Emulates a flash using a disk
```
Reworked

kica-z commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+properties:
+  size:
+    type: int
+    description: flash capacity in bits
```
Yes the emulated capacity. I adjusted the description accordingly. It was in bits since the jedec driver as well as the atmel did it that way. I rewrote it since it does not seem to be a convention.

kica-z commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+    type: string
+    description: the name of the underlying disk to be used
+    required: true
+  disk-sector-size:
+    type: int
+    description: the sector size in byte of the underlying disk to be used
+    required: true
+  disk-offset:
+    type: int
+    description: >
+      offset on the disk in bytes. If the offset is 10 the 10th byte will
+      be the first one used.
+    default: 0
```
True I could get the sector size at runtime which I already do but it is needed at compile-time in order to properly size the buffer `buf_##index`. Do you have a better proposal how to do this?

Why is disk-offset problematic? Or what exactly is problematic here?

jfischer-no replied:
The information is already there, there must not be an other property to describe the same, please take a look at DT_INST_PROP_BY_PHANDLE, ping me if you need help with that.

kica-z replied:
@jfischer-no I would appreciate some help here. I don't understand where this would already be defined since this is the offset for the driver in order to enable multiple partitions/flash instances on the same underlying disk.
 
 This enables the user to define multiple flash partitions on a disk by specifying their offset and size.

jfischer-no replied:
I mean disk-sector-size, I can try it out in the next few days and post it here.

jfischer-no replied:
@kica-z For the disk-name and disk-sector-size, see https://github.com/jfischer-no/zephyr/commit/0ad08a83ac43cc26209334453bfe38535355930a.patch, you can squash it into your commit an drop my signed-off.

kica-z replied:
@jfischer-no Very nice patch thanks for that 👍

Unfortunately it does not seem to work with `sdmmc-disk` as far as I could tell since no `sector-size` and `disk-name` devicetree properties are present. If that should be the case I don't think we should include the patch since it would narrow down the possible list of underlying disks.

I may however have made a mistake during testing which is unrelated to the changes in the patch. @jfischer-no what do you think? Should it work with the mmc disk node?

# Setup:

## Devicetree
```
&spi3 {
  status = "okay";
  cs-gpios = <&gpio0 5 GPIO_ACTIVE_LOW>;
  pinctrl-0 = <&spi3_pinctrl>;
  pinctrl-names = "default";

  sdhc0: sdhc@0 {
      compatible = "zephyr,sdhc-spi-slot";
      reg = <0>;
      status = "okay";
      pwr-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
      mmc0: mmc0 {                                                        <----- This was used as phandle
          compatible = "zephyr,sdmmc-disk";
          status = "okay";
      };
      spi-max-frequency = <8000000>;
  };
};
```

## Error
`error: 'DT_N_S_soc_S_peripheral_50000000_S_spi_b000_S_sdhc_0_S_mmc0_P_disk_name' undeclared here (not in a function); did you mean 'DT_N_S_soc_S_peripheral_50000000_S_spi_b000_S_sdhc_0_S_mmc0_P_status'?`

jfischer-no replied:
> @jfischer-no Very nice patch thanks for that 👍
> 
> Unfortunately it does not seem to work with `sdmmc-disk` as far as I could tell since no `sector-size` and `disk-name` devicetree properties are present. If that should be the case I don't think we should include the patch since it would narrow down the possible list of underlying disks.

That is unfortunately an issue in SDMMC driver, I spoke about this at ZDS2023, multi-instance support is broken for sdmmc drivers, someone needs to fix it.

kica-z replied:
Oh ok, that is unfortunate. Is there an issue open for this already?
I would propose not to include the patch until this has been fixed since sdmmc is the disk driver thats most relevant for this PR.

kica-z commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+    default: 0xFF
+  page-size:
+    type: int
+    description: size of a flash page in bytes
```
Adjusted

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+struct flash_disk_dev_config {
+	int flash_size;
+	struct flash_parameters flash_parameters;
+	char disk_name[Z_DEVICE_MAX_NAME_LEN];
```
You are righ. I adjusted it accordingly.

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+	int flash_size;
+	struct flash_parameters flash_parameters;
+	char disk_name[Z_DEVICE_MAX_NAME_LEN];
+	int disk_offset;
```
You are right. I replaced this.

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+};
+
+struct flash_disk_dev_config {
+	int flash_size;
```
You are right. I replaced this.

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+	char disk_name[Z_DEVICE_MAX_NAME_LEN];
+	int disk_offset;
+
+	IF_ENABLED(CONFIG_FLASH_PAGE_LAYOUT, (struct flash_pages_layout flash_pages_layout;))
```
Adjusted

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+};
+
+static void acquire_device(const struct device *dev);
+static void release_device(const struct device *dev);
```
Inlined the functions

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+	struct flash_disk_dev_data *data = dev->data;
+	const struct flash_disk_dev_config *cfg = dev->config;
+
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
```
I don't know if this is the case right now but should this be possible in the future the driver would be ready.

jfischer-no replied:
It is not, and this future is very unlikely.

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,312 @@
+static void acquire_device(const struct device *dev);
+static void release_device(const struct device *dev);
+
+static int flash_disk_init(const struct device *dev)
```
I moved the function to the bottom

kica-z commented on dts/bindings/flash_controller/zephyr,disk-flash.yaml:
```diff
@@ -0,0 +1,51 @@
+  erase-value:
+    type: int
+    description: byte value which should be used for erasing (range 0-255)
+    default: 0xFF
```
Since two people now told me it is not needed and I also don't see a use case I will remove the dt property and set it to 0.

kica-z commented:
@jfischer-no @carlescufi What is the status here? Is there something I am missing before a new review can be done?

kica-z commented on drivers/flash/Kconfig.disk:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2023 Endress+Hauser AG
+# SPDX-License-Identifier: Apache-2.0
+
+config FLASH_DRIVER_DISK
```
Renamed for better consistency

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,358 @@
+	uint32_t disk_sector_size;
+};
+
+struct flash_disk_dev_config {
```
Adjusted

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,358 @@
+	const struct flash_disk_dev_config *cfg;
+	uint32_t write_cursor = 0;
+
+	LOG_DBG("%s called", __func__);
```
Removed

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,358 @@
+{
+	int ret = 0;
+	char *dst_buf = (char *)dst;
+	struct flash_disk_dev_data *data;
```
I moved the pointer assignment as mentioned here.

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,358 @@
+static int flash_disk_read(const struct device *dev, off_t offset, void *dst, size_t len)
+{
+	int ret = 0;
+	char *dst_buf = (char *)dst;
```
Adjusted

kica-z commented on drivers/flash/flash_disk.c:
```diff
@@ -0,0 +1,358 @@
+		memcpy(&dst_buf[write_cursor], &data->sector_buf[read_padding], len_write);
+		write_cursor += len_write;
+	}
+release_device:
```
Adjusted

kica-z commented on samples/drivers/flash_disk/src/main.c:
```diff
@@ -0,0 +1,51 @@
+	}
+
+	return 0;
+}
```
I removed the sample and added overlays to work with the `flash_shell` sample.
-> Can be build by running: `west build -b qemu_x86 samples/drivers/flash_shell/ -DDTC_OVERLAY_FILE=disk_flash.overlay -DOVERLAY_CONFIG=disk_flash.conf`

kica-z commented on samples/drivers/flash_disk/prj.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_MINIMAL_LIBC=y
+
+# Enable flash on disk
+CONFIG_FLASH_DRIVER_DISK=y
```
Removed

kica-z commented on samples/drivers/flash_disk/app.overlay:
```diff
@@ -0,0 +1,25 @@
+/ {
+	ramdisk0 {
+		compatible = "zephyr,ram-disk";
+		disk-name = "DISK1";
```
Renamed

kica-z commented:
@jfischer-no Except for the point you mentioned with the `offset` everything is adjusted now. With the `offset` I would appreciate some support since there seems to be a bit of a misunderstanding.

de-nordic commented on drivers/flash/disk_flash.c:
```diff
@@ -0,0 +1,344 @@
+	}
+
+	while (read_cursor < len) {
+		uint32_t write_cursor = cfg->disk_offset + offset + read_cursor;
+		uint32_t write_sector = write_cursor / data->disk_sector_size;
+		uint32_t write_padding = write_cursor % data->disk_sector_size;
+		uint32_t len_write = (len - read_cursor);
```
I do not think it is very efficient to keep recalculating these in loop.

kica-z replied:
@de-nordic 
I agree that this is not very efficient but I don't see a way around it since the driver allows arbitrary write/read sizes and does not need to be sector aligned with the disk.

I understood you believe, that read/write/erase granularity must be equal to the underlying disk device. The idea of this flash driver is to offer a read/write/erase granularity of a single byte on top of a disk with arbitrary sector sizes. So operations typically:
-read sector,
-update sector,
-write sector,
-loop

clamattia replied:
But unless done this way, how would you simulate a flash with potentially different properties then underlying device?

Would you prefer, that `flash_disk` inherits properties from underlying device? If so, it would probably no longer be suitable for our use-case.
I see the following possibilities:
1. be at maximum permissive granularity (1 byte, current proposal)
2. be user configurable (with Kconfig?)
3. same as underlying device (severely limiting the usefulness of it to emulate a device with different properties on top of another)

Of course with 1. it is expected, that caller calls with big junks also. With 3. it is not possible to use as secondary partition for `mcuboot` if underlying device has different properties to main flash. Which would nullify the main benefit of the `flash_disk` module.

Or am I missing something? Can primary and secondary partition for mcuboot have different properties?
If yes, then of course, the story is entirely different.

Thanks for the quick responses and instructing reviews.

clamattia replied:
Or did I completely miss your point and you suggestions is just to improve simply the code to:
- handle initial (potentially misaligned) sector
- handle most sectors with single call to `disk_access
- handle final (potentially misaligned) sector

clamattia replied:
If I understood correctly this is exactly the Raison d'être of `flash_disk`. It provides a non-block-device interface to a part of a block device.

de-nordic replied:
The problem is that even flash drivers have something that is called "write-block-size" for ea reason. If you try to write size below that value you will end up damaging flash faster than expected by endurance characteristic of a device.
The problem we have here is that, for example, MCUboot will try to write some flags and it will do at 8 byte pace, which means that each time this happens the sector will be rewritten.
In case when MCUboot writes down status of swap operation it will write mark for each page it managed to swap, which means that the sector will be re-written as many times as the status is updated.
I am not sure how to solve this, but I do not think that locking the driver with 1 byte alignment is good idea.
If somebody picks the driver and starts to write info to SD card by doing writes by that alignment, the person will damage the SD card in unexpectedly short time - which means that we then have Zephyr issue logged.

kica-z replied:
@de-nordic Why would/does McuBoot use a 8byte pace when it could write any size possible?
Anyway in order for MCUBoot to work proberly the flash page size needs to be the same as the page size of the primary flash partition. Therefore I have added a DTS property again to be able to adjust this freely. The code still can handle any size and will write as much bytes as possible at once in order to not wear out the underlying disk.

de-nordic replied:
> @de-nordic Why would/does McuBoot use a 8byte pace when it could write any size possible? Anyway in order for MCUBoot to work proberly the flash page size needs to be the same as the page size of the primary flash partition. Therefore I have added a DTS property again to be able to adjust this freely. The code still can handle any size and will write as much bytes as possible at once in order to not wear out the underlying disk.

Looks like design requires 4:
https://github.com/mcu-tools/mcuboot/blob/c43a20fd193dc445faa9b7f0d2ed77759a6aca8f/docs/design.md?plain=1#L468-L474

Zephyr uses at least 8:
https://github.com/mcu-tools/mcuboot/blob/c43a20fd193dc445faa9b7f0d2ed77759a6aca8f/boot/bootutil/include/bootutil/bootutil_public.h#L88-L91

sure about the flash page size?
https://github.com/mcu-tools/mcuboot/blob/c43a20fd193dc445faa9b7f0d2ed77759a6aca8f/docs/design.md?plain=1#L184-L195

kica-z replied:
@de-nordic 

The primary and the secondary image need the same amount of sectors otherwise mcuboot fails at startup with the error message:
`W: Cannot upgrade: not a compatible amount of sectors`
But we are not using the scratch partition instead we are using the `Swap without using scratch`<

As far as I understood the problem is the writing of the partition header which needs to be written with every sector that is swapped. As I understood it there is nothing that can be done about this. But wouldn't the same problem occur with a flash since only a whole page can be written? 

BTW
I adjusted the driver in order to better emulate the write/erase behaviour by using a bitwise AND. I also added the write-block-size as a dts attribute which per default gets set to the sector size of the underlying this. This way the user explicitly has to set a lower block-size himself hopefully knowing the implication on sd-card-lifetime.

clamattia replied:
Thank you for elaborating your concerns, I think, that I understand better now. Did I understand your concerns correclty, that:
1. The trailer containing meta-information about the slot is written at the end of each slot. So there is a trailer at the end of the secondary slot on the flash_disk. Did I understand correctly, that this would mean, that whereas on a flash the trailer is erased once and then progressively written the flash-disk would need to read-update-write the sector containing the trailer with each status update which concerns you because of too much wear on disks without wear-leveling?
2. You are worried, that the write of a sector is not an atomic operation, such that it might end up being half-written for example in case of power-loss? Leading to possible damage to the system.

> There is another issue that there is very small chance that you can break the status update when record is added, but when you break it while write block happens, what will happen is that there will not longer be record of any swap happening and, at the same time, you will have two damaged slots (partially swapped).

Are you refering to the fact, that swapping might fail when using `swap-using-move` on an image that is too big (100% of available space) because it can not move down the primary imgage block-wise when there is no free block? Whereas this can not happen when using `swap-using-scratch`? Or is there another reason for recommending `swap using scratch`?

Thanks

Laczen replied:
Both with `swap-using-move` and `swap-using-scratch` the result of a block move is written to flash. When using a disk and the write of the last sector fails you could loose all information about the process (all information is on one disk sector), while on flash only the last block move information might be incorrect. If all information is lost there is no way to continue the swap and you are left with an unrecoverable system.

Laczen replied:
Yes, the strategy does not matter.

kica-z replied:
@de-nordic @Laczen 

Shouldn't the danger of sector write fail be lowered since when not written entirely the sd-card should still contain a partially correct sector? Or how is the status of the swap tracked? By settings bytes to 0?

Also I do not currently see any mitigation for the two problems brought up in this thread. Neither for the problem with the high disk usage due to sector overwrites nor for the problem with partial writes. Do you see any mitigation possibilites here or what would/should be the way forward?

de-nordic commented on drivers/flash/disk_flash.c:
```diff
@@ -0,0 +1,344 @@
+			data->disk_sector_size);
+	}
+
+	disk_size = data->disk_sector_size * data->disk_total_sector_cnt;
```
`size_t` may not be enough to hold the value. On 32 bit platform that is up to 4G. Is it possible that larger devices would be used?

kica-z replied:
You are right. There is absolutely a possibility that large disks could be used since this driver supports partitioning. I adjusted this.

de-nordic commented on drivers/flash/disk_flash.c:
```diff
@@ -0,0 +1,344 @@
+			len_write = data->sector_buf_size - write_padding;
+		}
+
+		memset(&data->sector_buf[write_padding],
+			cfg->flash_parameters.erase_value,
+			len_write);
```
You could do that once, before entering the loop.

kica-z replied:
Since the erase supports arbitrary size and does not need to be aligned with the sector of the disk partial sectors need to be erased. Therefore I don't currently see a possibility to do this outside the loop.

de-nordic commented on drivers/flash/disk_flash.c:
```diff
@@ -0,0 +1,344 @@
+	}
+
+	while (read_cursor < size) {
+		uint32_t write_cursor = cfg->disk_offset + offset + read_cursor;
+		uint32_t write_sector = write_cursor / data->disk_sector_size;
+		uint32_t write_padding = write_cursor % data->disk_sector_size;
+		uint32_t len_write = (size - read_cursor);
```
I guess that both cursors will be moved by sector size or multiple of that? Looks like too much arithmetic for adding sector size.

kica-z replied:
Not necessarily moved by sector size since the driver is not dependent on the underlying disk sector size partial sectors may be written or read. As such I think that much arithmetic is needed.

de-nordic commented on drivers/flash/disk_flash.c:
```diff
@@ -0,0 +1,344 @@
+		.flash_size = DT_INST_PROP(index, size),					\
+		.flash_parameters =								\
+		{										\
+			.write_block_size = 1,							\
```
I do not think this is good idea. If you try to loop writing data with such block size, what would happen is that you would probably get one sector erased and rewritten by number of bytes in sector. Or SD controller, if it is something super smart, will do number of sector reallocations equal to sector size in bytes.

kica-z replied:
I do not completely understand this comment but I had both write_block size and erase_size as DT property before but was told that this should not be necessary:
> Why would this need to be configurable? The underlying disk controller should always support 1 byte block writes

As I mentioned before this driver aims to be as independent from the underlying disk block size as possible and does currently support arbitrary read/write/erase sizes.

de-nordic commented on drivers/flash/disk_flash.c:
```diff
@@ -0,0 +1,344 @@
+			(.fpl =									\
+				{								\
+					.pages_count = DT_INST_PROP(index, size),		\
+					.pages_size = 1						\
```
Erase at least by sector size.

kica-z replied:
Is there a reason to make this driver dependent on the disk sector size? This driver currently supports arbitrary write/read/erase sizes independent from the underlying disk.

de-nordic commented on drivers/flash/disk_flash.c:
```diff
@@ -0,0 +1,344 @@
+		uint32_t read_padding = read_cursor % data->disk_sector_size;
+		uint32_t len_write = (len - write_cursor);
+
+		if (read_sector < 0 || read_sector >= data->disk_total_sector_cnt) {
+			ret = -EINVAL;
+			goto release_device;
+		}
```
Shouldn't the range be tested before entering the loop?

kica-z replied:
You are right. This check actually is not even needed anymore. With the `check_boundary` function we check that the read/write is within the defined flash partition and in `disk_flash_init` we check that the partition is within the underlying disk.

de-nordic commented on drivers/flash/disk_flash.c:
```diff
@@ -0,0 +1,344 @@
+		uint32_t write_padding = write_cursor % data->disk_sector_size;
+		uint32_t len_write = (len - read_cursor);
+
+		if (write_sector < 0 || write_sector >= data->disk_total_sector_cnt) {
+			LOG_ERR("Sector out of bounds");
+			ret = -EINVAL;
+			goto release_device;
+		}
```
Shouldn't the range be tested before entering the loop?

kica-z replied:
You are right. This check actually is not even needed anymore. With the `check_boundary` function we check that the read/write is within the defined flash partition and in `disk_flash_init` we check that the partition is within the underlying disk.

kica-z commented:
Added page-size as DTS property again since it is needed for mcuboot to be able to match the primary flash page size. Also defined the size in bits again since nrf-partition-manager gets confused otherwise (also done like this in other flash drivers).

jfischer-no commented:
> Added page-size as DTS property again since it is needed for mcuboot to be able to match the primary flash page size. Also defined the size in bits again since nrf-partition-manager gets confused otherwise (also done like this in other flash drivers).

Neither MCUBoot nor nrf-partition-manager (?) are Zephyr projects; there should be nothing specific to them in a Zephyr RTOS driver.

kica-z commented:
> > Added page-size as DTS property again since it is needed for mcuboot to be able to match the primary flash page size. Also defined the size in bits again since nrf-partition-manager gets confused otherwise (also done like this in other flash drivers).
> 
> Neither MCUBoot nor nrf-partition-manager (?) are Zephyr projects; there should be nothing specific to them in a Zephyr RTOS driver.

@jfischer-no I merely stated these softwares as the ones who showed me the errors the driver still was having. Defining disk size in bits is done in many other flash drivers as well so I merely adjusted the driver to better fit in with those drivers. Also being able to define the page size is a valuable additon for a flash-emulator no matter if used with mcuboot or not.
